makeflow/src/dag_node.c

```diff
@@ -220,6 +249,209 @@ void dag_node_add_target_file(struct dag_node *n, const char *filename, char *re
 	target->created_by = n;
 }
 
+void dag_node_prepare_node_size(struct dag_node *n)
+{
+	struct dag_file *f;
+	struct dag_node *s;
+	n->source_size = 0;
+	n->target_size = 0;
+	list_first_item(n->source_files);
+	while((f = list_next_item(n->source_files))){
+		if(dag_file_should_exist(f))
+			n->source_size += f->file_size;
+		else
+			n->source_size += f->est_size;
+	}
+	list_first_item(n->target_files);
+	while((f = list_next_item(n->target_files))){
+		if(dag_file_should_exist(f))
+			n->target_size += f->file_size;
+		else
+			n->target_size += f->est_size;
+	}
+
+	set_first_element(n->descendants);
+	while((s = set_next_element(n->descendants))){
+		dag_node_prepare_node_size(s);
```

Comment:
Note this function does not need to be recursive. It will be more efficient if you transverse the d->nodes list.